[feature] Automatically open threads for messages in specific channels

rajpiskala opened the pull request:
### Purpose of pull request

**Related Issue**: resolves #157

- Automatically opens a new thread for career questions
- Automatically opens a new thread for valid resume review submissions
  - Checks if a resume review submission has an image or PDF, otherwise deletes it

### Pull request checklist

- [x] Branch and PR are named correctly
- [x] Updated relevant documentation
- [x] Tested with a tester bot
- [x] Wrote unit tests for changes

NOTE: Please do not merge this before merging into these two PRs
1. Upgrade Discord.js to v13: #161 
2. Small PR to fix linting not working on Windows: #163 

### Screenshots
Career Questions:
![career-questions](https://user-images.githubusercontent.com/57082175/164338817-8a1f1081-15e1-4f1b-940e-d074982abbdc.png)

Resume Review:
![resume-review](https://user-images.githubusercontent.com/57082175/164338984-aa0a224b-b730-4cfc-ab38-3abb36e36579.gif)

### Testing instructions
1. Run unit tests using `yarn test`
2. Set up locally and designate two channels, one as a career question channel and one as a resume review channel to test manually

rajpiskala committed:
Upgrade Discord.js to v13 and bump up Jest version

rajpiskala committed:
Fix broken test after upgrade

rajpiskala committed:
Bump up Node version everywhere

rajpiskala committed:
Fix linting error

rajpiskala committed:
Fix ESLint and allow pre-commit hook to run on Windows

rajpiskala committed:
Support new channel IDs

rajpiskala committed:
Update ESLint to prevent yarn lint / yarn format conflict

rajpiskala committed:
Automatically open threads for career questions and resume review

rajpiskala committed:
Add unit tests for valid attachment

rajpiskala committed:
Add unit tests for valid attachment

Zacheriah commented on src/bots/monitor/index.js:
```diff
@@ -0,0 +1,44 @@
+  // Run routine based on id
+  switch (message.channelId) {
+    // Career questions channel
+    case get('CAREER_QUESTIONS_CHANNEL_ID'):
```
Consider adding a case for when a person replies to another message at the top-level instead of in a thread. (as a reply instead of as a standalone message). Also consider returning the message they sent in the dm you send them.

rajpiskala replied:
That's a really smart idea, will do both. Thanks!

Zacheriah commented on src/bots/monitor/index.js:
```diff
@@ -0,0 +1,44 @@
+
+      // Verify that all attachments are in valid format for a resume
+      if (!message.attachments.every((file) => attachmentIsAllowed(file))) {
+        author.send(defaultStrings.invalidFormat);
```
Consider sending the author of the invalid message a copy of their message when you delete it so they can repost it easily with the correct format.

rajpiskala replied:
That would save a lot of frustration. Will add that in the next revision, thanks for mentioning!

payne911 commented on src/bots/monitor/index.js:
```diff
@@ -0,0 +1,44 @@
+
+    // Resume review channel
+    case get('RESUME_REVIEW_CHANNEL_ID'):
+      // Verify a resume is attached to message
+      if (message.attachments.size === 0) {
+        author.send(defaultStrings.noResume);
+        message.delete();
+        return;
+      }
+
+      // Verify that all attachments are in valid format for a resume
+      if (!message.attachments.every((file) => attachmentIsAllowed(file))) {
+        author.send(defaultStrings.invalidFormat);
+        message.delete();
+        return;
+      }
```
The DM should include the text the user posted and which got deleted. This way, if they had written a massive wall-of-text, it's not lost.

rajpiskala replied:
Great idea, I'll add that to the next revision

payne911 reviewed (changes requested):
I love the demo GIF! I like your style.

Zacheriah reviewed (changes requested):
Looks great! I added some feedback on some lines I thought you could expand your implementation.

rajpiskala committed:
Resend message to user, delete any top-level replies

rajpiskala commented:
Made a new revision with the following changes:
- Replies to a top-level comment/thread will be auto-deleted
- Deleted replies will have their message contents AND attachments sent to the user as a direct message
  - Added code to handle the edge case where the content length is >2000 characters

Here's a gif demonstrating the new features:
![thread-create-rev-1](https://user-images.githubusercontent.com/57082175/164548739-8ba2de05-e683-4631-beed-9b6c704956ec.gif)

payne911 commented on .env.template:
```diff
@@ -4,4 +4,7 @@ ENV=testing
 GUILD_ID=1234567890
 MOD_CHANNEL_ID=0987654321
 DEV_CHANNEL_ID=0987654321
+CAREER_QUESTIONS_CHANNEL_ID=0987654321
+RESUME_REVIEW_CHANNEL_ID=0987654321
+SHARE_YOUR_CONTENT_CHANNEL_ID=0987654321
```
As discussed on Discord, I believe the auto-thread feature should also be applied to `#share-your-content` and `#hiring`

rajpiskala replied:
That's a lot more specific. Renamed the PR to that.

Will add both channels to the code in the next revision.

nikgil replied:
Because it's fairly dynamic I think it would be better to have a database entry of like a 1 col table where you can add/remove channels by command. Never know if we plan to add or remove channels from this practice

payne911 reviewed (changes requested):
Side-note: I would rename the PR
> [feature] Automatically open threads for messages in specific channels

rdeaton commented on .husky/pre-commit:
```diff
@@ -1,4 +1,5 @@
 #!/bin/sh
 . "$(dirname $0)/_/husky.sh"
 
-yarn format && yarn lint
```
This is a breaking change. yarn format can now fail, and the pre-commit script still passes. Probably best to add `set -x` to this script.

crabbycrabby commented on .husky/pre-commit:
```diff
@@ -1,4 +1,5 @@
 #!/bin/sh
 . "$(dirname $0)/_/husky.sh"
 
-yarn format && yarn lint
```
This is a breaking change, as previously yarn format failing would cause pre-commit to fail, and it does not anymore. Probably want `set -x` on verification scripts like this.

rajpiskala replied:
That's a good catch. I tested `yarn format && yarn lint` on Windows PowerShell just now and it seems to be working now while it wasn't earlier. I've added it #163, will need to drop the old commit from this pull request once #161 and #163 are merged.

rajpiskala committed:
Updated environment to add hiring and share-your-content channels

rajpiskala committed:
Add thread automatically opening for hiring and share-your-content channels

rajpiskala committed:
Add unit tests after supporting new channels

rajpiskala commented:
Made a new revision adding support for hiring and share your question channels along with some unit tests. Verified that the current functionality for career questions and resume review still works:
![image](https://user-images.githubusercontent.com/57082175/164951195-762b0c14-5099-44a0-a46b-ccced813cd58.png)

payne911 reviewed (approved):
The `yarn` comment from @crabbycrabby is beyond me, so I can't speak for that.

But the rest looks good to me. (Although my revision is very much on the surface since I am mostly a Java dev.)

I'm submitting an Approval so that my previous "Request changes" doesn't block this PR.

nikgil commented:
Can you try pull and merge with development branch?

nikgil commented on src/bots/monitor/index.js:
```diff
@@ -0,0 +1,85 @@
+        if (channelName === null) return;
+
+        // Do not allow top-level messages as replies
+        if (isMessageReply(message)) {
+          author.send(defaultStrings.invalidReply(channelName));
+          deleteAndReturnMessage(message);
+          return;
+        }
```
Need bypass for mods.

nikgil commented on src/bots/monitor/index.js:
```diff
@@ -0,0 +1,85 @@
+      break;
+
+    // Resume review channel
+    case getChannelIdForName(CHANNEL_NAMES.RESUME_REVIEW):
```
I get the reasoning behind all this, but referring to my more dynamic suggestion I think this is a single breakaway case that can be dealt with like this,

nikgil reviewed (commented):
Overall fine imo. Just needs the mod bypass stuff. Also the dynamic suggestion is more something for a future diff if the time comes to it. Also I merged your last diff in so this may present conflicts weith dev branch.
